core/storage/feedback/gae_models.py

```diff
@@ -420,6 +421,400 @@ def get_multi(cls, user_id, thread_ids):
+    # instances from old model to the new one. To be removed after migration.
+    last_updated = ndb.DateTimeProperty()
+
+    @classmethod
+    def _generate_id(cls, thread_id, message_id):
+        """Generates full message ID given the thread ID and message ID.
+
+        Args:
+            thread_id: str. Thread ID of the thread to which the message
+                belongs.
+            message_id: int. Message ID of the message.
+
+        Returns:
+            str. Full message ID.
+        """
+        return '.'.join([thread_id, str(message_id)])
+
+    @property
+    def entity_id(self):
+        """Returns the entity_id corresponding to this thread instance.
+
+        Returns:
+            str. The entity_id.
+        """
+        return self.id.split('.')[1]
+
+    @property
+    def entity_type(self):
+        """Returns the entity_type corresponding to this thread instance.
+
+        Returns:
+            str. The entity_type.
+        """
+        return self.id.split('.')[0]
+
+
+    def get_thread_subject(self):
+        """Returns the subject of the thread corresponding to this
+        GeneralFeedbackMessageModel instance.
+
+        Returns:
+            str. The subject of the thread.
+        """
+        return GeneralFeedbackThreadModel.get_by_id(self.thread_id).subject
+
+    @classmethod
+    def create(cls, thread_id, message_id):
+        """Creates a new FeedbackMessageModel entry.
+
+        Args:
+            thread_id: str. ID of the thread.
+            message_id: int. ID of the message.
+
+        Returns:
+            FeedbackMessageModel. Instance of the new FeedbackMessageModel
+                entry.
+
+        Raises:
+            Exception: A message with the same ID already exists
+                in the given thread.
+        """
+        instance_id = cls._generate_id(thread_id, message_id)
+        if cls.get_by_id(instance_id):
+            raise Exception('Feedback message ID conflict on create.')
+        return cls(id=instance_id)
+
+    @classmethod
+    def get(cls, thread_id, message_id, strict=True):
+        """Gets the FeedbackMessageModel entry for the given ID. Raises an
+        error if no undeleted message with the given ID is found and
+        strict == True.
+
+        Args:
+            thread_id: str. ID of the thread.
+            message_id: int. ID of the message.
+            strict: bool. Whether to raise an error if no FeedbackMessageModel
+                entry is found for the given IDs.
+
+        Returns:
+            GeneralFeedbackMessageModel or None. If strict == False and no
+                undeleted message with the given message_id exists in the
+                datastore, then returns None. Otherwise, returns the
+                GeneralFeedbackMessageModel instance that corresponds to the
+                given ID.
+
+        Raises:
+            EntityNotFoundError: If strict == True and message ID is not valid
```

Comment:
EntityNotFoundError: strict == True and either (i) the message ID is not valid, or (ii) the message is marked as deleted.

(Otherwise, hard to parse as (A and B) or C, vs A and (B or C).)

Reply:
Done